### ISSUE TITLE
test: Add integration tests for Autodiff module [P1]

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#628]
## Overview

This issue tracks integration tests for the **Autodiff** module. P1 (High Priority).

Parent Issue: #615

## Current Status
- **Coverage:** 0%
- **Source Files:** 5 files
- **Test Files:** None

## Module Location
`src/Autodiff/`

### Autodiff Classes
- GradientTape
- ComputationNode
- TensorOperations
- NeuralNetworkDerivatives
- GradientCheckpointing

### 1. GradientTape Tests
- [ ] Test gradient recording
- [ ] Test gradient computation
- [ ] Test nested tapes
- [ ] Test persistent tape
- [ ] Compare against PyTorch autograd

### 2. ComputationNode Tests
- [ ] Test forward pass
- [ ] Test backward pass
- [ ] Test node graph construction
- [ ] Test topological ordering

### 3. TensorOperations Tests
- [ ] Test differentiable operations
- [ ] Test gradient flow through ops
- [ ] Verify chain rule application

### 4. Derivative Tests
- [ ] Test neural network gradients
- [ ] Compare against numerical gradients
- [ ] Test higher-order derivatives

### 5. Gradient Checkpointing Tests
- [ ] Test memory savings
- [ ] Verify gradient correctness
- [ ] Test checkpoint boundaries

### Numerical Gradient Check
```
numerical_grad = (f(x + h) - f(x - h)) / (2h)
```
Compare against computed gradients.

### Test Functions
- f(x) = x^2 (simple)
- f(x) = sin(x) (transcendental)
- f(x, y) = x*y (multivariate)
- Neural network layers

## Edge Cases
- [ ] Zero gradients
- [ ] Very large gradients
- [ ] Discontinuous functions
- [ ] Non-differentiable points

## Acceptance Criteria
- [ ] All 5 classes tested
- [ ] Gradients match numerical computation
- [ ] Compare against PyTorch autograd
- [ ] Checkpointing verified
- [ ] 80%+ coverage

## References
- PyTorch autograd: https://pytorch.org/docs/stable/autograd.html
- JAX autodiff: https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html
